test/Microsoft.AspNet.Authentication.Test/OpenIdConnect/OpenIdConnectHandlerTests.cs

```diff
@@ -0,0 +1,647 @@
+                {
+                    SecurityTokenReceived = (notification) =>
+                    {
+                        notification.HandleResponse();
+                        return Task.FromResult<object>(null);
+                    }
+                };
+        }
+
+        private static void SecurityTokenReceivedSkippedOptions(OpenIdConnectAuthenticationOptions options)
+        {
+            options.StateDataFormat = new AuthenticationPropertiesFormater();
+            options.Notifications =
+                new OpenIdConnectAuthenticationNotifications
+                {
+                    SecurityTokenReceived = (notification) =>
+                    {
+                        notification.SkipToNextMiddleware();
+                        return Task.FromResult<object>(null);
+                    }
+                };
+        }
+
+        private static void SecurityTokenValidatedHandledOptions(OpenIdConnectAuthenticationOptions options)
+        {
+            options.StateDataFormat = new AuthenticationPropertiesFormater();
+            options.Notifications =
+                new OpenIdConnectAuthenticationNotifications
+                {
+                    SecurityTokenValidated = (notification) =>
+                    {
+                        notification.HandleResponse();
+                        return Task.FromResult<object>(null);
+                    }
+                };
+        }
+
+        private static void SecurityTokenValidatedSkippedOptions(OpenIdConnectAuthenticationOptions options)
+        {
+            options.StateDataFormat = new AuthenticationPropertiesFormater();
+            options.Notifications =
+                new OpenIdConnectAuthenticationNotifications
+                {
+                    SecurityTokenValidated = (notification) =>
+                    {
+                        notification.SkipToNextMiddleware();
+                        return Task.FromResult<object>(null);
+                    }
+                };
+        }
+
+        private static void DefaultOptions(OpenIdConnectAuthenticationOptions options)
+        {
+            options.StateDataFormat = new AuthenticationPropertiesFormater();
+        }
+
+        #endregion
+
+        private static TestServer CreateServer(Action<OpenIdConnectAuthenticationOptions> configureOptions, string scheme)
+        {
+            return TestServer.Create(app =>
+            {
+                app.UseOpenIdConnectAuthenticationPublic(configureOptions);
+                app.Use(async (context, next) =>
+                {
+                    await next();
+                });
+            },
+            services =>
+            {
+                services.AddDataProtection();
+            });
+        }
+    }
+
+    /// <summary>
+    /// Extension specifies <see cref="OpenIdConnectAuthenticationMiddlewarePublic"/> as the middleware.
+    /// </summary>
+    public static class OpenIdConnectAuthenticationExtensions
```

Comment:
These unit tests seem to need quite a lot of helper classes. Is there a design change we can make to the runtime types to make them easier to test? That usually leads to better design anyway.